gqcp/include/QCMethod/RHF/RHFSCFEnvironment.hpp

```diff
@@ -0,0 +1,118 @@
+// This file is part of GQCG-gqcp.
+// 
+// Copyright (C) 2017-2019  the GQCG developers
+// 
+// GQCG-gqcp is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+// 
+// GQCG-gqcp is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Lesser General Public License for more details.
+// 
+// You should have received a copy of the GNU Lesser General Public License
+// along with GQCG-gqcp.  If not, see <http://www.gnu.org/licenses/>.
+// 
+#pragma once
+
+
+#include "Basis/TransformationMatrix.hpp"
+#include "Mathematical/Representation/QCMatrix.hpp"
+#include "Operator/SecondQuantized/SQHamiltonian.hpp"
+#include "Processing/RDM/OneRDM.hpp"
+#include "QCMethod/RHF/RHF.hpp"
+
+#include <Eigen/Dense>
+
+
+namespace GQCP {
+
+
+/**
+ *  An algorithm environment that can be used with standard RHF SCF solvers.
+ * 
+ *  @tparam _Scalar             the scalar type that is used for the coefficient matrix/expansion coefficients
+ */
+template <typename _Scalar>
+class RHFSCFEnvironment {
+public:
+    using Scalar = _Scalar;
+
+
+public:
+    size_t N;  // the total number of electrons
+
+    std::deque<double> electronic_energies;
```

Comment:
At the moment, yes, these queues keep getting larger and larger with every iteration.

It might be advantageous, indeed, to implement some kind of automatic memory optimization for algorithm environments, by examining the number of required values in the iteration cycle.